src/hardware/dac.rs

```diff
@@ -0,0 +1,279 @@
+    pub max_i_neg1_pin: PB5<Alternate<AF2>>,
+    pub max_i_neg2_pin: PC8<Alternate<AF2>>,
+    pub max_i_neg3_pin: PC9<Alternate<AF2>>,
+}
+
+type Pt0<T, S> = super::hal::pwm::Pwm<T, S, ComplementaryDisabled, ActiveHigh, ActiveHigh>;
+type Pt1<T, S> = super::hal::pwm::Pwm<T, S, ComplementaryImpossible, ActiveHigh, ActiveHigh>;
+
+pub struct Pwm {
+    max_v0: Pt0<TIM1, C1>,
+    max_v1: Pt0<TIM1, C2>,
+    max_v2: Pt0<TIM1, C3>,
+    max_v3: Pt1<TIM1, C4>,
+    max_i_pos0: Pt1<TIM4, C1>,
+    max_i_pos1: Pt1<TIM4, C2>,
+    max_i_pos2: Pt1<TIM4, C3>,
+    max_i_pos3: Pt1<TIM4, C4>,
+    max_i_neg0: Pt1<TIM3, C1>,
+    max_i_neg1: Pt1<TIM3, C2>,
+    max_i_neg2: Pt1<TIM3, C3>,
+    max_i_neg3: Pt1<TIM3, C4>,
+}
+
+impl Pwm {
+    pub fn new(
+        clocks: &CoreClocks,
+        tim_rcc: (rec::Tim1, rec::Tim3, rec::Tim4),
+        tim: (TIM1, TIM3, TIM4),
+        pins: PwmPins,
+    ) -> Pwm {
+        fn init_pwm_pin<P: PwmPin<Duty = u16>>(pin: &mut P) {
+            pin.set_duty(0);
+            pin.enable();
+        }
+
+        let (mut max_v0, mut max_v1, mut max_v2, mut max_v3) = tim.0.pwm(
+            (
+                pins.max_v0_pin,
+                pins.max_v1_pin,
+                pins.max_v2_pin,
+                pins.max_v3_pin,
+            ),
+            F_PWM.khz(),
+            tim_rcc.0,
+            clocks,
+        );
+        init_pwm_pin(&mut max_v0);
+        init_pwm_pin(&mut max_v1);
+        init_pwm_pin(&mut max_v2);
+        init_pwm_pin(&mut max_v3);
+
+        let (mut max_i_pos0, mut max_i_pos1, mut max_i_pos2, mut max_i_pos3) = tim.2.pwm(
+            (
+                pins.max_i_pos0_pin,
+                pins.max_i_pos1_pin,
+                pins.max_i_pos2_pin,
+                pins.max_i_pos3_pin,
+            ),
+            F_PWM.khz(),
+            tim_rcc.2,
+            clocks,
+        );
+        init_pwm_pin(&mut max_i_pos0);
+        init_pwm_pin(&mut max_i_pos1);
+        init_pwm_pin(&mut max_i_pos2);
+        init_pwm_pin(&mut max_i_pos3);
+
+        let (mut max_i_neg0, mut max_i_neg1, mut max_i_neg2, mut max_i_neg3) = tim.1.pwm(
+            (
+                pins.max_i_neg0_pin,
+                pins.max_i_neg1_pin,
+                pins.max_i_neg2_pin,
+                pins.max_i_neg3_pin,
+            ),
+            F_PWM.khz(),
+            tim_rcc.1,
+            clocks,
+        );
+        init_pwm_pin(&mut max_i_neg0);
+        init_pwm_pin(&mut max_i_neg1);
+        init_pwm_pin(&mut max_i_neg2);
+        init_pwm_pin(&mut max_i_neg3);
+
+        Pwm {
+            max_v0,
+            max_v1,
+            max_v2,
+            max_v3,
+            max_i_pos0,
+            max_i_pos1,
+            max_i_pos2,
+            max_i_pos3,
+            max_i_neg0,
+            max_i_neg1,
+            max_i_neg2,
+            max_i_neg3,
+        }
+    }
+
+    /// Set PWM to limit current.
+    pub fn set(&mut self, ch: Channel, lim: Limit, val: f32) {
```

Comment:
The timer channels are unfortunately not homogeneous types (and the hal doesn't provide much type-erasure). How'd you put them into an array. Also the match isn't that large, even if you could put them into an array, there would probably be more lines of code.

Reply:
You can create your own wrapper enum around inconsistent types and allow it to be indexable. That's exactly what we do in Booster (https://github.com/quartiq/booster/blob/develop/src/hardware/rf_channel.rs#L67-L90). Perhaps we could make something like that a generic tool...

Reply:
Ah. Right. That would be one way. My guess is that it would be quite a bit of code (and/or macros which impact readability a bit) though.
IMO type-erase/degraded channels would be useful (and closer to standard C coding style) but those are a job for the HAL.